### PR TITLE
[repo] dpy2 asynchronous cog loading changes

### PR DESCRIPTION
**Pull request by japandotorg**
So uhh the dpy2 wasn't working for my `dpy2_unstable` branch of red and then i figured that it was bc the extensions weren't awaited, this PR will address those changes for the `dpy2_unstable` branch of red, had to have it done bc phen-cogs being the most used red-cog at the moment

1. Have you tested the PR?
- yes i have, and it works just as fine now with the `dpy2_unstable` branch

### CODE REVIEW

File: connect4/__init__.py [hunk 1]
```diff
@@ -31,4 +31,4 @@
 
 def setup(bot):
```
Comment (sravan1946) [inline connect4/__init__.py:32]: need to make this async

**Review by sravan1946 — CHANGES_REQUESTED** [review]

File: connect4/__init__.py (hunk 1 above)
Comment (japandotorg) [reply connect4/__init__.py:32]: oh yeah forgot to do that, that's for addressing!

**Review by sravan1946 — APPROVED** [review]
LGTM

**Comment by phenom4n4n** [comment]
Thanks for this PR. In order for it to be fully complete, the cogs should take advantage of the benefits that come with asynchronous cog loading. Specifically, the following changes also need to be made:

- [ ] Cogs that have a `cog_unload` method that calls `asyncio.create_task` should be made asynchronous and `await` the coroutine rather than creating a task. For example, in `aki/aki.py` at line 65,
```
    def cog_unload(self):
        asyncio.create_task(self.session.close())
```
should become
```
    async def cog_unload(self):
        await self.session.close()
```
- [ ] Cogs that call an asynchronous method in the setup function should remove that line and use `cog_load` instead which is implicitly called instead. For example, in `baron/__init__.py` at line 39, `await cog.build_cache()` should be removed and `build_cache` should be renamed to `cog_load` at `baron/baron.py` at line 95.

If you have questions about making these changes, feel free to ask in a comment.

**Comment by japandotorg** [comment]
sure, thanks for informing me, i'll start addressing them as soon as i'm on my pc and yes i'll post a comment if i need any help or whatsoever

**Comment by japandotorg** [comment]
After checking all the files thoroughly i have awaited all the coroutines and all the possible stuff that needed to be made asynchronous ( where i felt it was the right thing to do ), and I guess this is it for now, let me know if anything else needs a change or whatsoever.

File: slashtags/core.py [hunk 2]
```diff
@@ -123,13 +123,13 @@ def cog_unload(self):
         except Exception as error:
             log.exception("An error occurred while unloading the cog.", exc_info=error)
 
-    def __unload(self):
+    async def __unload(self):
         try:
-            self.bot.remove_dev_env_value("st")
+            await self.bot.remove_dev_env_value("st")
```
Comment (sravan1946) [inline slashtags/core.py:128]: this shouldn't be awaited iirc

File: tags/core.py [hunk 3]
```diff
@@ -110,12 +110,12 @@ def cog_unload(self):
         except Exception as e:
             log.exception("An error occurred during cog unload.", exc_info=e)
 
-    def __unload(self):
-        self.bot.remove_dev_env_value("tags")
+    async def __unload(self):
+        await self.bot.remove_dev_env_value("tags")
```
Comment (sravan1946) [inline tags/core.py:114]: same here

**Review by sravan1946 — CHANGES_REQUESTED** [review]

File: baron/__init__.py [hunk 4]
```diff
@@ -35,5 +35,5 @@
 
 async def setup(bot: Red) -> None:
     cog = Baron(bot)
-    bot.add_cog(cog)
-    await cog.build_cache()
+    await bot.add_cog(cog)
+    await cog.cog_load()
```
Comment (sravan1946) [inline baron/__init__.py:39]: This line should be removed  and `build_cache` should be renamed to `cog_load` at `baron/baron.py` at line 95

**Review by sravan1946 — CHANGES_REQUESTED** [review]

File: baron/__init__.py (hunk 4 above)
Comment (japandotorg) [reply baron/__init__.py:39]: oh thanks for pointing out didn't notice that it should have been renamed on the `baron.py` file

File: webhook/__init__.py [hunk 5]
```diff
@@ -34,4 +34,4 @@
 async def setup(bot):
     cog = Webhook(bot)
     await cog.initialize()
```
Comment (phenom4n4n) [inline webhook/__init__.py:36]: This should be refactored to a `cog_unload` method.

**Review by phenom4n4n — CHANGES_REQUESTED** [review]
Please make the changes requested in this latest review and then load and unload all cogs to ensure they have no errors. Make sure to watch the console for any errors that are logged implicitly without a user-facing error.

**Comment by japandotorg** [comment]
alright ty for pointing out all the stuff, i'll change them right away and test them

**Comment by japandotorg** [comment]
i'll start testing all the changes now

**Comment by japandotorg** [comment]
error
so the TagScriptEngine lib needs changes to be made.

**Comment by japandotorg** [comment]
custom-ping-error

this was invoked when testing the customping cog, was working fine before making the `cog_load` change.

**Comment by japandotorg** [comment]
```py
AttributeError: 'Member' object has no attribute 'avatar_url'
Ignoring exception in command tree:
Traceback (most recent call last):
  File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 881, in wrapper
    await self.call(interaction)
  File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 1020, in call
    command, options = self._get_app_command_options(data)
  File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 911, in _get_app_command_options
    raise CommandNotFound(name, parents)
discord.app_commands.errors.CommandNotFound: Application command 'ping' not found
Ignoring exception in view <ConfirmationView timeout=60 children=2> for item <Button style=<ButtonStyle.secondary: 2> url=None disabled=True label='No' emoji=None row=None>:
Traceback (most recent call last):
  File "/root/env/lib/python3.9/site-packages/discord/ui/view.py", line 413, in _scheduled_task
    await item.callback(interaction)
  File "/root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/slashtags/views.py", line 108, in no
    await self.disable_all(button, interaction)
  File "/root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/slashtags/views.py", line 115, in disable_all
    await interaction.response.edit_message(view=self)
AttributeError: 'Button' object has no attribute 'response'
[08:26:05] CRITICAL [red.main] Caught unhandled exception in <Task finished name='discord-ui-view-timeout-2e45611ecbe546b2fc67c08cbc5c188a' coro=<ConfirmationView.on_timeout() done, defined at /root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/slashtags/views.py:117> exception=NotFound('404 Not Found (error code: 10008): Unknown Message')>:
Task exception was never retrieved
╭───────────────────────────────────────────────────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────────────────────────────────────────────────╮
│ /root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/slashtags/views.py:119 in on_timeout                                                                                                    │
│ ❱ 119         await self.message.edit(view=self)                                                                                                                                                            │
│ /root/env/lib/python3.9/site-packages/discord/message.py:1984 in edit                                                                                                                                       │
│ ❱ 1984         data = await self._state.http.edit_message(self.channel.id, self.id,                                                                                                                         │
│        params=params)                                                                                                                                                                                       │
│ /root/env/lib/python3.9/site-packages/discord/http.py:493 in request                                                                                                                                        │
│ ❱  493                             raise NotFound(response, data)                                                                                                                                           │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
NotFound: 404 Not Found (error code: 10008): Unknown Message
Ignoring exception in command tree:
Traceback (most recent call last):
  File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 881, in wrapper
    await self.call(interaction)
  File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 1020, in call
    command, options = self._get_app_command_options(data)
  File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 911, in _get_app_command_options
    raise CommandNotFound(name, parents)
discord.app_commands.errors.CommandNotFound: Application command 'hi' not found
```

this is the error for slashtags

**Review by japandotorg — COMMENTED** [review]
```py
[08:30:49] ERROR    [red] Exception in command 'permlock perms'
╭───────────────────────────────────────────────────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────────────────────────────────────────────────╮
│ /root/env/lib/python3.9/site-packages/discord/ext/commands/core.py:187 in wrapped                                                                                                                           │
│ ❱  187             ret = await coro(*args, **kwargs)                                                                                                                                                        │
│ /root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/permissionslocker/permissionslocker.py:111 in perms                                                                                     │
│ ❱ 111             f"I will now require these permissions on commands:\n{box(await                                                                                                                           │
│       self.humanize_perms(permissions, True), 'diff')}"                                                                                                                                                     │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
TypeError: object str can't be used in 'await' expression
```
this was the error for this cog which was easily fixed

**Review by japandotorg — COMMENTED** [review]
```py
Ignoring exception in view <AkiView timeout=60 children=8> for item <Button style=<ButtonStyle.success: 3> url=None disabled=False label='yes' emoji=None row=None>:
Traceback (most recent call last):
  File "/root/env/lib/python3.9/site-packages/discord/ui/view.py", line 413, in _scheduled_task
    await item.callback(interaction)
  File "/root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/aki/views.py", line 44, in yes
    await self.answer_question("yes", interaction)
  File "/root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/aki/views.py", line 87, in answer_question
    await self.send_current_question(interaction)
  File "/root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/aki/views.py", line 167, in send_current_question
    await self.edit(interaction)
  File "/root/.local/share/Red-DiscordBot/data/esdeath/cogs/CogManager/cogs/aki/views.py", line 156, in edit
    await interaction.message.edit(embed=self.current_question_embed(), view=self)
AttributeError: 'Button' object has no attribute 'message'
```

fixed the error for aki

**Review by japandotorg — COMMENTED** [review]
this still needs changes to be done here, I'll do them after sometime, I have some other things to take care of for now

**Comment by japandotorg** [comment]
Okay so after extensive use of my single braincell I've fixed whatever I could and tested each and every cog + the sub commands, and all that is left are the tags and slashtags cog ( TagScriptEngine lib related ) and the customping cog.

File: antihonde/__init__.py [hunk 6]
```diff
@@ -35,5 +35,5 @@
 
 async def setup(bot: Red):
     antihonde = AntiHonde(bot)
-    await antihonde.initialize()
-    bot.add_cog(antihonde)
+    await antihonde.cog_load()
```
Comment (sravan1946) [inline antihonde/__init__.py:38]: Remove this line and rename `initialize` in antihonde.py line.49 to `cog_load`

File: customping/customping.py [hunk 7]
```diff
@@ -221,5 +221,5 @@ async def setup(bot):
         bot.remove_command(old_ping.name)
 
     cog = CustomPing(bot)
-    await cog.initialize()
-    bot.add_cog(cog)
+    await cog.cog_load()
```
Comment (sravan1946) [inline customping/customping.py:224]: Remove this line and rename `initialize` in customping.py line.57 to `cog_load`

File: permissionslocker/__init__.py [hunk 8]
```diff
@@ -35,5 +35,5 @@
 
 async def setup(bot: Red) -> None:
     cog = PermissionsLocker(bot)
-    await cog.initialize()
-    bot.add_cog(cog)
+    await cog.cog_load()
```
Comment (sravan1946) [inline permissionslocker/__init__.py:38]: Remove this line and rename `initialize` in permissionslocker.py line.64 to `cog_load`

File: slashtags/__init__.py [hunk 9]
```diff
@@ -57,5 +57,5 @@
 async def setup(bot: Red):
     await validate_tagscriptengine(bot, tse_version)
     cog = SlashTags(bot)
-    await cog.pre_load()
-    bot.add_cog(cog)
+    await cog.cog_load()
```
Comment (sravan1946) [inline slashtags/__init__.py:60]: Remove this line and rename `pre_load` in core.py line.134 to `cog_load`

File: webhook/__init__.py [hunk 10]
```diff
@@ -33,5 +33,5 @@
 
 async def setup(bot):
     cog = Webhook(bot)
-    await cog.initialize()
-    bot.add_cog(cog)
+    await cog.cog_load()
```
Comment (sravan1946) [inline webhook/__init__.py:36]: Remove this line and rename `initialize` in webhook.py line.64 to `cog_load`

**Review by sravan1946 — CHANGES_REQUESTED** [review]

**Comment by japandotorg** [comment]
The dpy2 branch of TagScriptEngine??

**Comment by phenom4n4n** [comment]
> ```py
> Ignoring exception in command tree:
> 
> Traceback (most recent call last):
> 
>   File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 881, in wrapper
> 
>     await self.call(interaction)
> 
>   File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 1020, in call
> 
>     command, options = self._get_app_command_options(data)
> 
>   File "/root/env/lib/python3.9/site-packages/discord/app_commands/tree.py", line 911, in _get_app_command_options
> 
>     raise CommandNotFound(name, parents)
> 
> discord.app_commands.errors.CommandNotFound: Application command 'hi' not found
> 
> ```
> 
> 
> 
> this is the error for slashtags

Do slashtags still work when invoked? As far as I can tell, this error should only be logged implicitly but won't prevent slashtags from being invoked. I'm going to fix it at a later date.

**Comment by phenom4n4n** [comment]
> The dpy2 branch of TagScriptEngine??

Yeah, but that's unrelated to this PR.

**Comment by japandotorg** [comment]
> I forgot about this change, can you go through all view callbacks in the repo and change the signature from `(self, button: discord.ui.Button, interaction: discord.Interaction)` to `(self, interaction: discord.Interaction, button: discord.ui.Button)?

Sure, i'll address them and do it right away.

File: banchart/banchart.py [hunk 11]
```diff
@@ -67,7 +67,7 @@ async def red_delete_data_for_user(self, *, requester: str, user_id: int) -> Non
     @staticmethod
     async def get_ban_limit(ctx: commands.Context, limit: int) -> Tuple[int, list]:
         await ctx.trigger_typing()
-        bans = await ctx.guild.bans()
+        bans = [bans async for bans in ctx.guild.bans()]
```
[the remark this answers is not in the document]
Comment (japandotorg) [reply banchart/banchart.py:70]: alright, but just know, if i revert it, it won't work anymore. I changed it bc I was getting an error when trying to invoke the banchart command.

**Review by phenom4n4n — APPROVED** [review]

**Comment by phenom4n4n** [comment]
Thanks for making the changes I asked for throughout this PR!

**Pull request merged**